File: src/Systems/Rust/bridge.rs
trait Device {
    fn power_on(&self) -> &'static str;
    fn mute(&self) -> &'static str;
}

struct TvDevice;
impl Device for TvDevice {
    fn power_on(&self) -> &'static str { "TV:on" }
    fn mute(&self) -> &'static str { "TV:muted" }
}

struct RadioDevice;
impl Device for RadioDevice {
    fn power_on(&self) -> &'static str { "Radio:on" }
    fn mute(&self) -> &'static str { "Radio:muted" }
}

struct BasicRemote<'a, D: Device> { device: &'a D }
impl<D: Device> BasicRemote<'_, D> {
    fn activate(&self) -> &'static str { self.device.power_on() }
}

struct MuteRemote<'a, D: Device> { device: &'a D }
impl<D: Device> MuteRemote<'_, D> {
    fn activate(&self) -> &'static str { self.device.mute() }
}

fn main() {
    let tv = TvDevice;
    let radio = RadioDevice;
    println!("basic-tv={}", BasicRemote { device: &tv }.activate());
    println!("basic-radio={}", BasicRemote { device: &radio }.activate());
    println!("mute-tv={}", MuteRemote { device: &tv }.activate());
    println!("mute-radio={}", MuteRemote { device: &radio }.activate());
}

File: src/Systems/Swift/bridge.swift
protocol Device {
    func powerOn() -> String
    func mute() -> String
}

struct TvDevice: Device {
    func powerOn() -> String { "TV:on" }
    func mute() -> String { "TV:muted" }
}

struct RadioDevice: Device {
    func powerOn() -> String { "Radio:on" }
    func mute() -> String { "Radio:muted" }
}

protocol RemoteControl {
    func activate() -> String
}

struct BasicRemote: RemoteControl {
    let device: any Device
    func activate() -> String { device.powerOn() }
}

struct MuteRemote: RemoteControl {
    let device: any Device
    func activate() -> String { device.mute() }
}

let tv = TvDevice()
let radio = RadioDevice()
print("basic-tv=\(BasicRemote(device: tv).activate())")
print("basic-radio=\(BasicRemote(device: radio).activate())")
print("mute-tv=\(MuteRemote(device: tv).activate())")
print("mute-radio=\(MuteRemote(device: radio).activate())")
